### PR TITLE
[develop] Add process bufr task.

### LINKED ISSUE #486: BUFR Processing tasks are needed for RRFS
## Description
RRFS runs several executables included with rrfs-utl as a preparation of existing RAP bufr files. This task corresponds to a task in RRFS_dev1 named process_bufr. It does the following:

- Run `process_Lightning.exe` on files like `*lghtng.tm00.bufr_d`
- Run `process_larccld.exe` on files like `*lgycld.tm00.bufr_d`
- Run `process_metarcld.exe` on files like `*prepbufr.tm00 `

## Solution
Use the RRFS_dev1 version of this task as a guide to a solution. Include any necessary default configuration settings, add j-jobs, and ex-scripts, and hook them into the workflow. 

These tasks have no dependencies other than data retrieval, so should be able to stand alone as a full workflow without other components.

## Requirements**
- All contributions must be machine independent and follow the style and guidelines of the SRW App.
- The contributed scripts must be tested, and new tests should be contributed for SRW. These WE2E tests can be short and only include data retrieval and running the task.

## Acceptance Criteria (Definition of Done)
Matches answers obtained in RRFS. Please reach out to someone at NOAA GSL to help with this test case.

### PR DESCRIPTION
**Pull request by danielabdi-noaa**
## DESCRIPTION OF CHANGES: 

Add `process_bufr` task from `RRFS_dev1`.
Note that there is no workflow so this PR can not be run independently until then.
However, it did successfully run for both spinup/production cycles in PR #540.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## TESTS CONDUCTED: 

- [X] hera.intel
  Run fundamental tests. Also run conus 3km test case from RRFS_dev1 successfully as shown in the run in PR #540 
  
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [ ] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [ ] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

## ISSUE: 

- #486 

## CHECKLIST

- [X] My code follows the style guidelines in the Contributor's Guide
- [ ] I have performed a self-review of my own code using the Code Reviewer's Guide
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## LABELS (optional): 

A Code Manager needs to add the following labels to this PR: 
- [ ] Work In Progress
- [ ] bug
- [ ] enhancement
- [ ] documentation
- [ ] release
- [ ] high priority
- [ ] run_ci
- [ ] run_we2e_fundamental_tests
- [ ] run_we2e_comprehensive_tests
- [ ] Needs Cheyenne test 
- [ ] Needs Jet test 
- [ ] Needs Hera test 
- [ ] Needs Orion test 
- [ ] help wanted

### CODE REVIEW

File: scripts/exregional_process_bufr.sh
Comment (MichaelLueken) [inline scripts/exregional_process_bufr.sh]: observation rather than obseration.

File: scripts/exregional_process_bufr.sh
Comment (MichaelLueken) [inline scripts/exregional_process_bufr.sh]: For lightning bufr files, `run_lightning=false` is set first, while for NASA LaRC and METAR cloud bufr files, `run_cloud` and `run_metar` are set just before `if [ -r "${obs_file}" ]; then`.  While I'm okay with this, would it be best to move `run_lightning=false` to line 146 to make all three entries more similar?

**Review by MichaelLueken — COMMENTED** [review]
@danielabdi-noaa Thanks for creating a separate PR to integrate the changes you made for processing bufr files from PR #540!  Overall, everything looks good to me!

I have noted one minor spelling issue in `scripts/exregional_process_bufr.sh`.  I also left a comment in the same script about the placement of `run_lightning` (suggested moving it from line 144 to line 146 to bring it into alignment with `run_cloud` and `run_metar` further in the script).

**Review by MichaelLueken — APPROVED** [review]
@danielabdi-noaa Thank you very much for addressing my comments!  Approving now.

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [inline scripts/exregional_process_bufr.sh]: We probably don't want to include logic for the support of an additional system at this time, and limit it only to RRFS.

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [inline scripts/exregional_process_bufr.sh]: I think this type of information should live in the appropriate machine file, moving us away from any machine case statements.

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [inline scripts/exregional_process_bufr.sh]: Is this a current supported grid?

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [inline scripts/exregional_process_bufr.sh]: If this line of the if statement was moved to line 243, it will be more efficient since the executable isn't meant to run.

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [inline scripts/exregional_process_bufr.sh]: The same efficiency statement goes here, too.

File: jobs/JREGIONAL_PROCESS_BUFR
Comment (christinaholtNOAA) [inline jobs/JREGIONAL_PROCESS_BUFR:7]: It would be helpful to include what type of process this does, and which executables are called. Perhaps @hu5970 could shed some light on exactly what the processing is?

**Review by christinaholtNOAA — COMMENTED** [review]
Thanks for breaking this out, @danielabdi-noaa.

File: ush/config_defaults.yaml
Comment (MichaelLueken) [inline ush/config_defaults.yaml]: located instead of loacated

**Review by MichaelLueken — COMMENTED** [review]
@danielabdi-noaa With the update from hash [10574e9](https://github.com/ufs-community/ufs-srweather-app/pull/636/commits/10574e93377f26501c2e44830a5cfd4783e418af), a minor spelling error was introduced in `ush/config_defaults.yaml` (loacated instead of located).

File: scripts/exregional_process_bufr.sh
Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: I am afraid this is inevitable. For example it would be awkward  to put the below code in a machine file. Even if it is possible, I don't think it is a good idea to put such a large piece of code in a machine file.

https://github.com/ufs-community/ufs-srweather-app/blob/2fd2c059388e9d6d33911ca5b1ced14e757703dc/scripts/exregional_run_anal.sh#L197-L267

File: scripts/exregional_process_bufr.sh
Comment (MichaelLueken) [inline scripts/exregional_process_bufr.sh]: Looking through PR #540, it doesn't look like you needed to add the `GSD_RAP13km` grid parameters to `ush/predef_grid_params.yaml` in order for you to successfully run the RRFS.  Would it be best to remove the GSD_RAP13km grid (remove lines 237, 238, 239, 240, and 243), or should we begin supporting the `GSD_RAP13km` grid, as you currently have it here and in `ush/predef_grid_params.yaml`?

**Review by MichaelLueken — COMMENTED** [review]
@danielabdi-noaa While looking at PR #540, it doesn't look like the `GSD_RAP13km` was added to `ush/predef_grid_params.yaml` in order to make the workflow run.  Should we move forward with supporting this new grid, or should we go ahead and remove it from `scripts/exregional_process_bufr.sh `and `ush/predef_grid_params.yaml`?

File: scripts/exregional_process_bufr.sh
Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: @MichaelLueken Yes, the test case I run is the RRFS CONUS 3km so this grid was not needed for my run. However, RRFS adds in total 3 new predefined grids and they are used in other test cases. So I think it is best to have it even though it is not needed for this particular test case. I can move the addition of the predefined grid to this PR #647 to make reviewing of PRs for tasks easier. PR #647 will contain common variables for tasks, common predef grids etc. What do you think?

Comment (MichaelLueken) [reply scripts/exregional_process_bufr.sh]: @danielabdi-noaa Thank you very much for the explanation.  So, it sounds like there will be two more grids that will need to be added, in addition to `GSD_RAP13km`.  Since the idea behind PR #647 is to contain common variables and grids, I think it would be fine to update `ush/predef_grid_params.yaml` in PR #647.  I would suggest, after updating the predef grids, to leave a comment in the PRs that contain the new predef grids pointing to PR #647 so that the reviewers know where they can be found.

Comment (MichaelLueken) [reply scripts/exregional_process_bufr.sh]: @danielabdi-noaa Looking through old PRs, I see that @gsketefian opened PR [#458](https://github.com/ufs-community/regional_workflow/pull/458) in the regional_workflow repository that renamed `GSD_RAP13km` to `RRFS_NA_13km`.  I see this grid in `ush/predef_grid_params.yaml`.  Following this, `GSD_RAP13km` should be renamed to `RRFS_NA_13km` in this PR.

Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: @MichaelLueken I have replaced `GSD_RAP13km` with `RRFS_NA_13km`. It has some slight differences but given #458, it is best to keep SRW values. I will put the values from RRFS_dev1 for the `RRFS_NA_13km` grid in comments just for future reference in PR #647

File: scripts/exregional_process_bufr.sh
Comment (mkavulich) [reply scripts/exregional_process_bufr.sh]: I disagree that case statements are "inevitable". The example you point out here is largely identical code with just a few changes; it might require some refactoring but I can't imagine this couldn't be handled without a `case` statement. 

In the case of the current PR, it would be as simple as making `obsfileprefix` a variable read in from the machine file.

File: scripts/exregional_process_bufr.sh
Comment (mkavulich) [inline scripts/exregional_process_bufr.sh]: Not sure why there are so many spaces here

```suggestion
cp_vrfy ${fixgriddir}/fv3_grid_spec fv3sar_grid_spec.nc
```

File: scripts/exregional_process_bufr.sh
Comment (mkavulich) [inline scripts/exregional_process_bufr.sh]: `SUBH` doesn't seem to get set to anything except a blank string, do we know what this is supposed to do?

File: scripts/exregional_process_bufr.sh
Comment (mkavulich) [inline scripts/exregional_process_bufr.sh]: It looks like this should be `grid_type` based on the namelist below
```suggestion
#   grid_type     : grid type (background will be used in the analysis)
```

File: scripts/exregional_process_bufr.sh
Comment (mkavulich) [inline scripts/exregional_process_bufr.sh]: Same here
```suggestion
#   grid_type     : grid type (background will be used in the analysis)
```

File: scripts/exregional_process_bufr.sh
Comment (mkavulich) [inline scripts/exregional_process_bufr.sh]: Is this the only grid where these parameters should be used? Or is it resolution-based?

File: scripts/exregional_process_bufr.sh
Comment (mkavulich) [inline scripts/exregional_process_bufr.sh]: This indentation is inconsistent with the surrounding code
```suggestion
  PREP_STEP
  eval $RUN_CMD_UTILS ${exec_fp} ${REDIRECT_OUT_ERR} || print_err_msg "\
       Call to executable to run lightning process returned with nonzero exit code."
  POST_STEP
```

File: scripts/exregional_process_bufr.sh
Comment (mkavulich) [inline scripts/exregional_process_bufr.sh:265]:
```suggestion
```

File: scripts/exregional_process_bufr.sh
Comment (mkavulich) [inline scripts/exregional_process_bufr.sh]:
```suggestion
```

**Review by mkavulich — CHANGES_REQUESTED** [review]
A few questions and suggested changes

File: scripts/exregional_process_bufr.sh
Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: @mkavulich Note that for this PR, `obsfileprefix` is a function of a variable defined in the script itself namely `obs_source` for non-wcoss2 machines. That depends on the cycle hour HH. I am not saying it is not possible to do it, but is it worth it? The way I see it we will end up with some cryptic code embedded in a string, far from where it is used, so it will be harder to understand the code. It may be acceptable for something short and generic (used in most ex-scripts) such as `PRE_TASK_CMDS` that mostly sets `ulimit -a`, but I don't think it is a good idea to do this routinely for everything machine specific.

File: scripts/exregional_process_bufr.sh
Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: @mkavulich There used to be an RTMA code that sets `SUBH` to START_DATE but now that is gone (because we don't want to support RTMA), it is now always set to empty string.

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [inline scripts/exregional_process_bufr.sh]:
```suggestion
This is the ex-script for the task that runs bufr (cloud, metar, lightning) preprocessing
```

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [reply scripts/exregional_process_bufr.sh]:
```suggestion

```

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [reply scripts/exregional_process_bufr.sh]: I'm seeing now, too, that `OBSPATH` hasn't been defined in this PR. 

We don't yet have a task in the workflow to get the bufr observations for input to this task, but we do have some of the bufr observations locations defined in machine files already. Ensuring that the bufr files are retrieved appropriately via that avenue should remove the need for any of this logic in this script. The only thing this script will need is the location that SRW stages the files in the experiment directory once retrieve_data.py is run to fetch them.

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [inline scripts/exregional_process_bufr.sh]:
```suggestion
  Build rrfs_utl and rerun."
```

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [inline scripts/exregional_process_bufr.sh]:
```suggestion
  Build rrfs_utl and rerun."
```

File: scripts/exregional_process_bufr.sh
Comment (christinaholtNOAA) [inline scripts/exregional_process_bufr.sh]:
```suggestion
  Build rrfs_utl and rerun."
```

**Review by christinaholtNOAA — CHANGES_REQUESTED** [review]

File: scripts/exregional_process_bufr.sh
Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: `SUBH` is used later in two places to define `obs_file` in `t${HH}{SUBH}z` although it is empty for non-RTMA cases. I could remove it completely in all places or keep the definition as empty so that it would be easy to identify which lines should change in case RTMA is added back.

Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: I have removed SUBH completely.

File: jobs/JREGIONAL_PROCESS_BUFR
Comment (danielabdi-noaa) [reply jobs/JREGIONAL_PROCESS_BUFR:7]: I have added the description from the associated issue for this PR.

File: scripts/exregional_process_bufr.sh
Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: I have templated the path to be `OBSFILE_TEMPLATE` that contains the path and prefix. The default value covers most machines, for WCOSS2 specifically, it is defined in the corresponding machine file.
I have tested the task again in the bigger PR by incorporating these new changes.

Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: OBSPATH is defined in the high-priority PR #647 that this PR depends on. FIXgsi too, however I have mentioned these dependencies in the j-job description.

File: scripts/exregional_process_bufr.sh
Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: NOAA-GSL fork have added "RRFS_CONUS_13km" to the if-conditon since last time I updated the code. It looks like both are 13km resolution.

**Comment by danielabdi-noaa** [comment]
@mkavulich @christinaholtNOAA I believe i have addressed your concerns. Let me know if I missed something.

**Review by mkavulich — APPROVED** [review]
Thanks!

File: ush/config_defaults.yaml
Comment (christinaholtNOAA) [inline ush/config_defaults.yaml]: Should this one go in the process_bufr section? Is it needed by other tasks?

**Review by christinaholtNOAA — APPROVED** [review]
LGTM. Just one minor question below.

File: ush/config_defaults.yaml
Comment (danielabdi-noaa) [reply ush/config_defaults.yaml]: Moved.

**Comment by MichaelLueken** [comment]
@danielabdi-noaa The Jenkins tests on Cheyenne, Gaea, and Orion successfully passed.  Jenkins is unable to communicate with Jet, which led to the Jet tests being aborted.  The WE2E fundamental tests were manually run on Jet and all the tests successfully passed.  Is this work ready to be merged, or should it wait until PR #647 has been merged first (checking due to the dependency on PR #647 above)?  Since this work doesn't affect the FV3LAM_wfow.xml file, I can move forward with the merge.  Thanks!

File: scripts/exregional_process_bufr.sh
Comment (EdwardSnyder-NOAA) [inline scripts/exregional_process_bufr.sh]: The predefined grids are now located under $DOMAIN_PREGEN_BASEDIR and not fixgriddir. See PR #638 for guidance.

Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: Thanks! Fixed it now.

**Comment by danielabdi-noaa** [comment]
@MichaelLueken I think merging PR #647 first is better. Also the yaml workflow PR would necessitate changes in this PR too. Thanks!

File: scripts/exregional_process_bufr.sh
Comment (EdwardSnyder-NOAA) [inline scripts/exregional_process_bufr.sh]: @danielabdi-noaa just noticed that you are using the CDATE variable here. @christinaholtNOAA suggested to use the ${PDY}${cyc} variables instead as they are NCO-compliant variables. You can see this discussion in #644.

Also, take a look at the date variables below. Some of these variables aren't used in this script like `PREYYJJJHH`, `YYYYMMDD`, etc.

Comment (danielabdi-noaa) [reply scripts/exregional_process_bufr.sh]: Thanks fixed it now!

**Comment by christinaholtNOAA** [comment]
@MichaelLueken We discussed in our tag up yesterday that we would like to go ahead and get this PR through as one of our higher priorities. Let's not hold it waiting for #647 as we previously considered. Thanks!

**Comment by MichaelLueken** [comment]
@EdwardSnyder-NOAA Just checking to see if there are any other issues that you see with respect to moving forward with these changes.  I have gone ahead and relaunched the Jenkins tests.  So long as they continue to pass, then I will move forward with these changes.  Thanks!

**Comment by EdwardSnyder-NOAA** [comment]
> @EdwardSnyder-NOAA Just checking to see if there are any other issues that you see with respect to moving forward with these changes. I have gone ahead and relaunched the Jenkins tests. So long as they continue to pass, then I will move forward with these changes. Thanks!

No other issues I see. Proceed with testing and merging.

**Comment by venitahagerty** [comment]
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1264903189/20230412185015/ufs-srweather-app
Build was Successful
Rocoto jobs started
Experiment done: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
2023-04-12 19:33:47 +0000 :: hfe07 :: This cycle is complete: Success
Experiment failed: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
2023-04-12 19:46:09 +0000 :: hfe07 :: Task run_MET_GridStat_vx_REFC_mem000, jobid=43764373, in state DEAD (FAILED), ran for 17.0 seconds, exit status=256, try=1 (of 1)
Experiment failed: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
2023-04-12 19:46:09 +0000 :: hfe07 :: Task run_MET_GridStat_vx_RETOP_mem000, jobid=43764374, in state DEAD (FAILED), ran for 17.0 seconds, exit status=256, try=1 (of 1)
Experiment failed: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
2023-04-12 19:46:25 +0000 :: hfe07 :: Task run_MET_PcpCombine_fcst_APCP01h_mem000, jobid=43764384, in state DEAD (FAILED), ran for 15.0 seconds, exit status=256, try=2 (of 2)
Experiment done: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
2023-04-12 19:31:43 +0000 :: hfe07 :: This cycle is complete: Success
Experiment done: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
2023-04-12 19:28:01 +0000 :: hfe07 :: This cycle is complete: Success
Experiment done: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp_regional_plot
2023-04-12 19:29:02 +0000 :: hfe07 :: This cycle is complete: Success
Experiment done: pregen_grid_orog_sfc_climo
2023-04-12 19:27:15 +0000 :: hfe07 :: This cycle is complete: Success
Experiment done: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
2023-04-12 19:26:28 +0000 :: hfe07 :: This cycle is complete: Success
If test failed, please make changes and add the following label back:
ci-hera-intel-WE

**Comment by MichaelLueken** [comment]
The Jenkins tests have successfully completed.  For those who can access the Jenkins pipeline, it can be found - `https://jenkins.epic.oarcloud.noaa.gov/blue/organizations/jenkins/ufs-srweather-app%2Fpipeline/detail/PR-636/2/pipeline`.  both manual and ci-hera-intel-WE tests were run and both passed (with the exception of PcpCombine due to issue #688).  Moving forward with merging this work.

**Pull request merged**
